test-data/unit/check-redefine2.test

```diff
@@ -0,0 +1,1076 @@
+        return
+    reveal_type(x) # N: Revealed type is "builtins.str"
+
+[case testNewRedefineBreakAndContinue]
+# flags: --allow-redefinition-new --local-partial-types
+def b() -> None:
+    while int():
+        x = ""
+        if int():
+            x = 1
+            break
+        reveal_type(x) # N: Revealed type is "builtins.str"
+        x = None
+    reveal_type(x) # N: Revealed type is "Union[builtins.int, None]"
+
+def c() -> None:
+    x = 0
+    while int():
+        reveal_type(x) # N: Revealed type is "builtins.int" \
+                       # N: Revealed type is "Union[builtins.int, builtins.str, None]"
+        if int():
+            x = ""
+            continue
+        else:
+            x = None
+        reveal_type(x) # N: Revealed type is "None"
+    reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str, None]"
+
+[case testNewRedefineUnderscore]
+# flags: --allow-redefinition-new --local-partial-types
+def f() -> None:
+    if int():
+        _ = 0
+        reveal_type(_) # N: Revealed type is "builtins.int"
+    else:
+        _ = ""
+        reveal_type(_) # N: Revealed type is "builtins.str"
+    reveal_type(_) # N: Revealed type is "Union[builtins.int, builtins.str]"
+
+[case testNewRedefineWithStatement]
+# flags: --allow-redefinition-new --local-partial-types
+class C:
+    def __enter__(self) -> int: ...
+    def __exit__(self, x, y, z): ...
+class D:
+    def __enter__(self) -> str: ...
+    def __exit__(self, x, y, z): ...
+
+def f1() -> None:
+    with C() as x:
+        reveal_type(x) # N: Revealed type is "builtins.int"
+    with D() as x:
+        reveal_type(x) # N: Revealed type is "builtins.str"
+
+def f2() -> None:
+    if int():
+        with C() as x:
+            reveal_type(x) # N: Revealed type is "builtins.int"
+    else:
+        with D() as x:
+            reveal_type(x) # N: Revealed type is "builtins.str"
+    reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str]"
+
+[case testNewRedefineTryStatement]
+# flags: --allow-redefinition-new --local-partial-types
+class E(Exception): pass
+
+def g(): ...
+
+def f1() -> None:
+    try:
+        x = 1
+        g()
+        x = ""
+        reveal_type(x) # N: Revealed type is "builtins.str"
+    except RuntimeError as e:
+        reveal_type(e) # N: Revealed type is "builtins.RuntimeError"
+        reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str]"
+    except E as e:
+        reveal_type(e) # N: Revealed type is "__main__.E"
+        reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str]"
+    reveal_type(e) # N: Revealed type is "<Deleted 'e'>"
+    reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str]"
+
+def f2() -> None:
+    try:
+        x = 1
+        if int():
+            x = ""
+            return
+    except Exception:
+        # TODO: Too wide
+        reveal_type(x) # N: Revealed type is "Union[builtins.int, builtins.str]"
+    # TODO: Too wide
```

Comment:
Is this problem specific to try/except? If yes, it is fine to leave it for a follow-up PR.

Reply:
Actually the types are probably the best we can reasonably infer. I'll remove the comment (and update the test case slightly).